bindings/sundials4py/arkode/arkode.cpp

```diff
@@ -0,0 +1,244 @@
+                       nb::arg("arkode_mem"), nb::arg("postprocessstep").none());
+
+  BIND_ARKODE_CALLBACK(ARKodeSetPostprocessStageFn, ARKPostProcessFn,
+                       postprocessstagefn, arkode_postprocessstagefn_wrapper,
+                       nb::arg("arkode_mem"), nb::arg("postprocessstage").none());
+
+  BIND_ARKODE_CALLBACK(ARKodeSetStagePredictFn, ARKStagePredictFn,
+                       stagepredictfn, arkode_stagepredictfn_wrapper,
+                       nb::arg("arkode_mem"), nb::arg("stagepredict").none());
+
+  BIND_ARKODE_CALLBACK(ARKodeSetNlsRhsFn, ARKRhsFn, nlsfi,
+                       arkode_nlsrhsfn_wrapper, nb::arg("arkode_mem"),
+                       nb::arg("nls_fi").none());
+
+  BIND_ARKODE_CALLBACK(ARKodeSetJacFn, ARKLsJacFn, lsjacfn,
+                       arkode_lsjacfn_wrapper, nb::arg("arkode_mem"),
+                       nb::arg("jac").none());
+
+  BIND_ARKODE_CALLBACK(ARKodeSetMassFn, ARKLsMassFn, lsmassfn,
+                       arkode_lsmassfn_wrapper, nb::arg("arkode_mem"),
+                       nb::arg("mass").none());
+
+  BIND_ARKODE_CALLBACK2(ARKodeSetPreconditioner, ARKLsPrecSetupFn,
+                        lsprecsetupfn, arkode_lsprecsetupfn_wrapper,
+                        ARKLsPrecSolveFn, lsprecsolvefn,
+                        arkode_lsprecsolvefn_wrapper, nb::arg("arkode_mem"),
+                        nb::arg("psetup").none(), nb::arg("psolve").none());
+
+  BIND_ARKODE_CALLBACK2(ARKodeSetMassPreconditioner, ARKLsMassPrecSetupFn,
+                        lsmassprecsetupfn, arkode_lsmassprecsetupfn_wrapper,
+                        ARKLsMassPrecSolveFn, lsmassprecsolvefn,
+                        arkode_lsmassprecsolvefn_wrapper, nb::arg("arkode_mem"),
+                        nb::arg("psetup").none(), nb::arg("psolve").none());
+
+  BIND_ARKODE_CALLBACK2(ARKodeSetJacTimes, ARKLsJacTimesSetupFn,
+                        lsjactimessetupfn, arkode_lsjactimessetupfn_wrapper,
+                        ARKLsJacTimesVecFn, lsjactimesvecfn,
+                        arkode_lsjactimesvecfn_wrapper, nb::arg("arkode_mem"),
+                        nb::arg("jtsetup").none(), nb::arg("jtimes").none());
+
+  BIND_ARKODE_CALLBACK(ARKodeSetJacTimesRhsFn, ARKRhsFn, lsjacrhsfn,
+                       arkode_lsjacrhsfn_wrapper, nb::arg("arkode_mem"),
+                       nb::arg("jtimesRhsFn").none());
+
+  m.def(
+    "ARKodeSetMassTimes",
+    [](void* ark_mem,
+       std::function<std::remove_pointer_t<ARKLsMassTimesSetupFn>> msetup,
+       std::function<std::remove_pointer_t<ARKLsMassTimesVecFn>> mtimes)
+    {
+      auto fn_table                = get_arkode_fn_table(ark_mem);
+      fn_table->lsmasstimessetupfn = nb::cast(msetup);
+      fn_table->lsmasstimesvecfn   = nb::cast(mtimes);
+      return ARKodeSetMassTimes(ark_mem, &arkode_lsmasstimessetupfn_wrapper,
+                                &arkode_lsmasstimesvecfn_wrapper, nullptr);
```

Comment:
The macros will pass `nullptr` to the ARKode functions when needed. Should this (and similarly for the other explicitly wrapped functions) be something like
```suggestion
     if (msetup && mtimes) 
     {
        return ARKodeSetMassTimes(ark_mem, &arkode_lsmasstimessetupfn_wrapper,
                                  &arkode_lsmasstimesvecfn_wrapper, nullptr);
     }
     else if (msetup)
     {
        return ARKodeSetMassTimes(ark_mem, nullptr, &arkode_lsmasstimesvecfn_wrapper, nullptr);
     }
     else if (msolve)
     {
        return ARKodeSetMassTimes(ark_mem, &arkode_lsmasstimessetupfn_wrapper, nullptr, nullptr);
     }
     else
     {
        return ARKodeSetMassTimes(ark_mem, nullptr, nullptr, nullptr);
     }
```

Reply:
Yes, we should do this.